sites/all/modules/ding_unilogin/ding_unilogin.module

```diff
@@ -0,0 +1,342 @@
+               array('%time' => $_GET['timestamp'], '%user' => $_GET['user']),
+               WATCHDOG_WARNING);
+      drupal_set_message(t('Could not understand the reply from UNI•Login, so cannot log you in.'), 'error');
+    }
+    // Redirect to current_path() to clear most GET parameters.
+    $query = array();
+    if (isset($_GET['ding-unilogin-trigger']) && $success) {
+      // This should propagate if we logged in a user, else it doesn't matter
+      // (doesn't make sense to re-display the reservation popup, as it'll
+      // just trigger the login form again).
+      $query['ding-unilogin-trigger'] = $_GET['ding-unilogin-trigger'];
+    }
+    drupal_goto(current_path(), array('query' => $query));
+  }
+}
+
+/**
+ * URL to redirect to for UNI•Login.
+ */
+function ding_unilogin_url($path = NULL) {
+  $unilogin_id = variable_get('ding_unilogin_id', '');
+
+  // If no path given default to redirecting back to current path.
+  if (!$path && request_uri() != '/') {
+    $path = url(request_uri(), array('absolute' => TRUE));
+  }
+
+  if (!$path) {
+    return sprintf('https://sso.emu.dk/unilogin/login.cgi?id=%s', $unilogin_id);
+  }
+
+  $secret = variable_get('ding_unilogin_secret', '');
+  return sprintf('https://sso.emu.dk/unilogin/login.cgi?id=%s&path=%s&auth=%s',
+                 $unilogin_id,
+                 urlencode(base64_encode($path)),
+                 md5($path . $secret)
+  );
+}
+
+/**
+ * Check whether a user is logged in via UNI•Login.
+ */
+function ding_unilogin_authenticated($account = NULL) {
+  global $user;
+  if (!$account) {
+    $account = $user;
+  }
+
+  if (isset($account->data['unilogin']) && $account->data['unilogin']) {
+    return TRUE;
+  }
+  return FALSE;
+}
+
+/**
+ * Attemp to log in UNI•Login user.
+ *
+ * @param string $username
+ *   The UNI•Login user name.
+ *
+ * @return bool
+ *   Whether the user was logged in.
+ */
+function _ding_unilogin_login($username) {
+  $auth_service = _ding_unilogin_wservice('auth');
+  $access = FALSE;
+
+  try {
+    // Check that the user has license to access us.
+    $access = $auth_service->hasLicense(
+      $username,
+      variable_get('ding_unilogin_auth_context', ''),
+      variable_get('ding_unilogin_auth_project', '')
+    );
+
+    // Look up the retailer id we should use for the user.
+    if ($access) {
+      $info_service = _ding_unilogin_wservice('info');
+      $municipality = $info_service->getMunicipality($username);
+      $municipalities = variable_get('ding_unilogin_municipality_mapping', array());
+      if (isset($municipalities[$municipality])) {
+        $publizon_id = $municipalities[$municipality];
+
+        // Check that the retailer_id is one of the configured libraries.
+        $publizon_libraries = variable_get('publizon_libraries', array());
+        if (!isset($publizon_libraries[(int) $publizon_id])) {
+          $access = FALSE;
+        }
+      }
+      else {
+        $access = FALSE;
```

Comment:
No it's not. $access can be true in L244, while the if stement in L249  triggers this else, and sets it to false.

But the flow could use some work.